lib/screens/screens_by_alert/memcache.ex

```diff
@@ -1,29 +1,176 @@
+    _ = Task.Supervisor.start_child(TaskSupervisor, fn -> do_put(screen_id, alert_ids) end)
+
     :ok
   end
 
   @impl true
-  def get_screens_by_alert(alert_ids) do
-    Map.new(alert_ids, &{&1, []})
+  def get_screens_by_alert(alert_ids) when is_list(alert_ids) do
+    now = System.system_time(:second)
+    default_map = Map.new(alert_ids, &{&1, []})
+    cache_keys = Enum.map(alert_ids, &alert_key/1)
+
+    case Memcache.multi_get(@server, cache_keys) do
+      {:ok, cache_result_map} ->
+        found_items = Map.new(cache_result_map, &clean_up_alert_cache_item(&1, now))
+        Map.merge(default_map, found_items)
+
+      {:error, message} ->
+        Logger.warn("[get_screens_by_alert memcache error] message=\"#{message}\"")
+        # Should we return an error tuple instead of the default map?
+        default_map
+    end
   end
 
   @impl true
-  def get_screens_last_updated(screen_ids) do
-    Map.new(screen_ids, &{&1, 0})
+  def get_screens_last_updated(screen_ids) when is_list(screen_ids) do
+    default_map = Map.new(screen_ids, &{&1, 0})
+    cache_keys = Enum.map(screen_ids, &last_updated_key/1)
+
+    case Memcache.multi_get(@server, cache_keys) do
+      {:ok, cache_result_map} ->
+        found_items = Map.new(cache_result_map, &clean_up_screens_last_updated_cache_item/1)
+        Map.merge(default_map, found_items)
+
+      {:error, message} ->
+        Logger.warn("[get_screens_last_updated memcache error] message=\"#{message}\"")
+        # Should we return an error tuple instead of the default map?
+        default_map
+    end
+  end
+
+  defp do_put(screen_id, alert_ids) do
+    now = System.system_time(:second)
+
+    # Put last updated timestamp under screens_last_updated.<id>
+    set_result =
+      Memcache.set(@server, last_updated_key(screen_id), now, ttl: @screens_last_updated_ttl)
+
+    _ =
+      case set_result do
+        {:error, message} ->
+          Logger.warn(
+            "[put_data screens_last_updated memcache error] screen_id=#{screen_id} message=#{message}"
+          )
+
+        _ ->
+          :ok
+      end
+
+    Enum.each(alert_ids, &update_alert_key(&1, screen_id, now))
   end
+
+  # Creates or updates the cache item for the given alert ID with the given screen ID, retrying until success.
+  # In the process of updating existing screen IDs under the alert key, this also removes
+  # any screen IDs that have expired.
+  defp update_alert_key(alert_id, screen_id, now) do
+    key = alert_key(alert_id)
+    timestamped_screen_id = {screen_id, now}
+
+    cas_result =
+      Memcache.cas(
+        @server,
+        key,
+        fn timestamped_screen_ids ->
+          unexpired_ids =
+            Enum.reject(timestamped_screen_ids, fn {_id, timestamp} ->
+              timestamp + @screens_ttl < now
+            end)
+
+          [timestamped_screen_id | unexpired_ids]
+        end,
+        retry: true,
+        default: [timestamped_screen_id],
```

Comment:
When is the default used as a fallback by memcache?

Reply:
The default is used when the key does not already exist in the cache. This is basically how `Memcache.cas` works:
1. Check if the key that we want to update exists
2a. If it does, get its value, run the updater function on it to produce its new value, and set that value.
2b. If it does not, create the key using the default value instead.